Mathlib/RingTheory/Bialgebra/TensorProduct.lean

```diff
@@ -0,0 +1,178 @@
+  refine Bialgebra.mk' R (A ⊗[R] B) ?_ (fun {x y} => ?_) ?_ (fun {x y} => ?_) <;>
+  simp_all only [AlgHom.toLinearMap_apply] <;>
+  simp only [_root_.map_one, _root_.map_mul]
+
+namespace Bialgebra.TensorProduct
+
+variable {R A B C D : Type u} [CommRing R] [Ring A] [Ring B] [Ring C] [Ring D]
+    [Bialgebra R A] [Bialgebra R B] [Bialgebra R C] [Bialgebra R D]
+
+/-- The tensor product of two bialgebra morphisms as a bialgebra morphism. -/
+noncomputable def map (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    A ⊗[R] C →ₐc[R] B ⊗[R] D :=
+  { Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D),
+    Algebra.TensorProduct.map (f : A →ₐ[R] B) (g : C →ₐ[R] D) with }
+
+@[simp]
+theorem map_tmul (f : A →ₐc[R] B) (g : C →ₐc[R] D) (x : A) (y : C) :
+    map f g (x ⊗ₜ y) = f x ⊗ₜ g y :=
+  rfl
+
+@[simp]
+theorem map_toCoalgHom (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    map f g = Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D) := rfl
+
+@[simp]
+theorem map_toAlgHom (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    (map f g : A ⊗[R] C →ₐ[R] B ⊗[R] D) =
+      Algebra.TensorProduct.map (f : A →ₐ[R] B) (g : C →ₐ[R] D) :=
+  rfl
+
+variable (R A B C) in
+/-- The associator for tensor products of R-bialgebras, as a bialgebra equivalence. -/
+protected noncomputable def assoc :
+    (A ⊗[R] B) ⊗[R] C ≃ₐc[R] A ⊗[R] (B ⊗[R] C) :=
+  { Coalgebra.TensorProduct.assoc R A B C, Algebra.TensorProduct.assoc R A B C with }
+
+@[simp]
+theorem assoc_tmul (x : A) (y : B) (z : C) :
+    Bialgebra.TensorProduct.assoc R A B C ((x ⊗ₜ y) ⊗ₜ z) = x ⊗ₜ (y ⊗ₜ z) :=
+  rfl
+
+@[simp]
+theorem assoc_symm_tmul (x : A) (y : B) (z : C) :
+    (Bialgebra.TensorProduct.assoc R A B C).symm (x ⊗ₜ (y ⊗ₜ z)) = (x ⊗ₜ y) ⊗ₜ z :=
+  rfl
+
+@[simp]
+theorem assoc_toCoalgEquiv :
+    (Bialgebra.TensorProduct.assoc R A B C : _ ≃ₗc[R] _) =
+    Coalgebra.TensorProduct.assoc R A B C := rfl
+
+@[simp]
+theorem assoc_toAlgEquiv :
+    (Bialgebra.TensorProduct.assoc R A B C : _ ≃ₐ[R] _) =
+    Algebra.TensorProduct.assoc R A B C := rfl
+
+variable (R A) in
+/-- The base ring is a left identity for the tensor product of bialgebras, up to
+bialgebra equivalence. -/
+protected noncomputable def lid : R ⊗[R] A ≃ₐc[R] A :=
+  { Coalgebra.TensorProduct.lid R A, Algebra.TensorProduct.lid R A with }
+
+@[simp]
+theorem lid_toCoalgEquiv :
+    (Bialgebra.TensorProduct.lid R A : R ⊗[R] A ≃ₗc[R] A) = Coalgebra.TensorProduct.lid R A := rfl
+
+@[simp]
+theorem lid_toAlgEquiv :
+    (Bialgebra.TensorProduct.lid R A : R ⊗[R] A ≃ₐ[R] A) = Algebra.TensorProduct.lid R A := rfl
+
+@[simp]
+theorem lid_tmul (r : R) (a : A) : Bialgebra.TensorProduct.lid R A (r ⊗ₜ a) = r • a := rfl
+
+@[simp]
+theorem lid_symm_apply (a : A) : (Bialgebra.TensorProduct.lid R A).symm a = 1 ⊗ₜ a := rfl
+
+theorem coalgebra_rid_eq_algebra_rid_apply (x) :
+    Coalgebra.TensorProduct.rid R A x = Algebra.TensorProduct.rid R R A x :=
+  congr($((TensorProduct.AlgebraTensorModule.rid_eq_rid R A).symm) x)
```

Comment:
The fact that this isn't `rfl` makes me think that one of the other two definitions is suboptimal? Did you already investigate how easy this is to fix? A `TODO: make this defeq, which would involve [...]` or similar is enough for this PR, as is "note: not possible to make defeq because [...`]

Reply:
It's not `rfl` because I didn't make `Coalgebra.TensorProduct.rid` heterobasic like the RHS is. I guess that's another drawback of my categorical approach, it's probably not the most amenable to adding heterobasic definitions. I've added a `TODO` comment.